### PR TITLE
Enforce structured task format and prohibit brittle patterns

### PR DESCRIPTION
**Pull request by JMBattista**
## Summary
- **Primary outcome**: Establish and enforce a mandatory structured task format (bold imperative title + behavioral description + acceptance criteria bullets) across all task authoring prompts. Prohibit brittle patterns that break under codebase drift: line-number references, exact code prescriptions, test mechanics, restated acceptance scenarios, and exact error strings.
- **Notable behaviour changes**: Task decomposition agents (smithy.slice, smithy.reconcile-slices) and the task auditing checklist now validate against the structured format and prohibit overspecified, implementation-prescriptive tasks. The reconcile-slices agent now reformats non-compliant tasks during cleanup.
- **Follow-up work deferred**: None.

## Context
Task authoring in the Smithy agentic system has suffered from overspecification and brittleness. Tasks that embed exact line numbers, copy-paste code, test stub configurations, or restated acceptance scenario content become stale as the codebase evolves between planning and implementation. Fresh-context dispatch (where each task runs in isolation with no memory of prior tasks) amplifies this problem: a task that prescribes "update line 68" or "mock the database with `{ id: 1 }`" fails when the implement agent cannot see the context that made those prescriptions valid.

This change codifies best practices into mandatory structure and explicit prohibitions, ensuring tasks remain resilient to codebase drift and leverage the spec file and contracts that the implement agent already receives.

### Key Changes

1. **Structured Task Format (Mandatory)**
   - Title: imperative verb phrase, max 12 words
   - Description: 2–3 sentences naming target file/module and outcome; reference acceptance scenarios by ID (e.g., "AS 2.1")
   - Acceptance criteria: 3–7 bulleted observable invariants
   - Total length: 50–150 words per task

2. **Reference, Don't Restate**
   - Tasks reference acceptance scenarios by ID instead of copying their content
   - Implement agent receives spec file path and reads scenarios directly
   - Contracts and data model sections referenced by name, not copied

3. **Prohibitions (Explicit)**
   - No standalone test tasks (TDD protocol handles testing)
   - No research or file-reading tasks (fresh context, no memory between tasks)
   - No verification tasks (Forge runs validation post-slice)
   - No baked-in test expectations (TDD determines test approach)
   - No line-number references or exact code (drift-prone)
   - **No test mechanics** (stub configs, mock patterns, assertion structures)
   - **No exact error strings or function signatures** (reference the spec instead)

4. **Auditing & Cleanup**
   - smithy.cut.prompt: expanded Task Scoping audit criterion to check format compliance and prohibitions
   - smithy.slice.prompt: added format requirements and prohibitions to task generation guidelines
   - smithy.reconcile-slices.prompt: added cleanup step to reformat non-compliant tasks and strip prescriptive detail
   - audit-checklist-tasks.md: updated to reflect new format and prohibitions

### Impacted Modules
- `src/templates/agent-skills/commands/smithy.cut.prompt` — audit checklist and task format documentation
- `src/templates/agent-skills/agents/smithy.slice.prompt` — task generation guidelines and output format
- `src/templates/agent-skills/agents/smithy.reconcile-slices.prompt` — cleanup and format validation
- `src/templates/agent-skills/snippets/audit-checklist-tasks.md` — audit checklist reference

### Before/After Contrast
The diff includes a detailed before/after example showing how an overspecified task (~300 words, embedding exact error strings, stub configs, line numbers) is rewritten into the structured format (~80 words, referencing the spec and acceptance scenarios).

## Risks & Mitigations

| Risk | Mitigation |
|------|-----------|
| Agents may initially produce non-compliant tasks during decomposition | smithy.reconcile-slices now reformats tasks during cleanup; parent agent reviews and can reject non-compliant slices |
| Existing task files may not follow the new format | This is a

https://claude.ai/code/session_01JtAFwdsqTYBBL8ZCGfAgCq

### CODE REVIEW

File: src/templates/agent-skills/commands/smithy.cut.prompt [hunk 1]
```diff
@@ -44,7 +44,7 @@ Use the **smithy-refine** sub-agent. Pass it:
   | **Task Completeness** | Are tasks within each slice sufficient to achieve the slice goal? Are there missing steps (tests, docs, validation)? |
   | **FR Traceability** | Does every slice trace to at least one FR or acceptance scenario from the user story? Are any FRs unaddressed? |
   | **Dependency Order** | Is the recommended implementation sequence logical? Would reordering reduce risk or unblock parallel work? |
-  | **Task Scoping** | Do tasks describe *what* to accomplish (not *how* with exact code)? Are there standalone test tasks, file-reading tasks, verification tasks, or line-number references that would break fresh-context dispatch? |
+  | **Task Scoping** | Do tasks follow the structured format (bold title + behavioral description + acceptance criteria bullets)? Are any tasks over 150 words? Do tasks reference acceptance scenarios by ID rather than restating their content? Are test mechanics absent (no stub configs, mock patterns, assertion structures)? Are there standalone test tasks, file-reading tasks, verification tasks, line-number references, or exact error strings that would break fresh-context dispatch or create brittleness? |
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt]: Fixed in b6a9395: Added "exact code prescriptions, exact function signatures" to the Task Scoping audit criteria. The audit question now covers all prohibited brittle patterns.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/snippets/audit-checklist-tasks.md [hunk 2]
```diff
@@ -6,6 +6,6 @@
...
 | **Testability** | Is it clear how each slice should be tested? Are integration test concerns addressed? |
 | **Edge Case Coverage** | Are boundary conditions, error paths, and failure modes covered in the tasks? |
-| **Task Scoping** | Do tasks describe *what* to accomplish without prescribing exact code, line numbers, or copy-paste replacements? Are there standalone test tasks (should be part of TDD), file-reading/research tasks (break fresh-context dispatch), verification tasks (handled by forge), or baked-in test expectations (pre-empt TDD)? |
+| **Task Scoping** | Do tasks follow the structured format (bold title + behavioral description + acceptance criteria bullets)? Are any tasks over 150 words? Do tasks reference acceptance scenarios by ID rather than restating their content? Are test mechanics absent (no stub configs, mock patterns, assertion structures, exact error strings)? Are there standalone test tasks (should be part of TDD), file-reading/research tasks (break fresh-context dispatch), verification tasks (handled by forge), or baked-in test expectations (pre-empt TDD)? |
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/snippets/audit-checklist-tasks.md]: Fixed in b6a9395: Added "exact function signatures" to the prohibited patterns list in the audit-checklist-tasks.md Task Scoping row.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt (hunk 1 above)
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt]: Fixed in b6a9395: Same fix as the sibling comment — added "exact code prescriptions, exact function signatures" to the Phase 0 Task Scoping audit criteria.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt [hunk 3]
```diff
@@ -277,32 +282,105 @@ Each task is dispatched to a **fresh sub-agent** (smithy-implement) with no
...
+
+> - [ ] **Extend `checkSpawnDependencies()` to validate all spawn preconditions**
+>
+>   Add a `baseImage` parameter to the function in `src/deps.ts`. Expand it to
+>   validate all preconditions from US2 acceptance scenarios 2.1–2.5, returning
+>   the first failure encountered.
+>
+>   _Acceptance criteria:_
+>   - Existing finder-availability gate preserved
+>   - Git-missing error matches AS 2.1 wording (current message diverges)
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt]: Fixed in b6a9395: Removed the parenthetical "(current message diverges)" — the criterion now reads simply "Git-missing error matches AS 2.1 wording", which is timeless.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt [hunk 4]
```diff
@@ -277,32 +282,105 @@ Each task is dispatched to a **fresh sub-agent** (smithy-implement) with no
...
+  - ...
+```
+
+- **Title**: imperative verb phrase, max 12 words. Scannable at a glance.
+- **Description**: 2–3 sentences. States WHAT changes and WHERE.
+- **Acceptance criteria**: 3–7 bullets. Observable invariants — what the
+  implement agent verifies via TDD.
+- **Total length**: aim for 50–100 words. Tasks over 150 words are almost
+  certainly overspecified — split them or trim prescriptive detail.
+
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt:309]: Not addressed: The structured task format applies to the cut/forge pipeline (spec → tasks → implement), not to strike. Strike is a lightweight single-session workflow where the plan and implementation happen in one context — there's no fresh-context dispatch, so the drift-resilience rationale that drives these rules doesn't apply. Expanding to strike would be a separate change with different design considerations.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/agents/smithy.reconcile-slices.prompt [hunk 5]
```diff
@@ -96,9 +96,22 @@ the task:
...
-parent agent can see what was cleaned up.
+- **No test mechanics.** If a task prescribes stub configurations, mock
+  objects, assertion patterns, or test helper modifications, strip them and
+  express the underlying requirement as an acceptance criterion.
+- **No restated acceptance scenarios.** If a task copies acceptance scenario
+  content from the spec instead of referencing it by ID (e.g., "AS 2.1"),
+  replace with a reference. The implement agent receives the spec file path.
+- **No exact error strings or function signatures.** If a task embeds the
+  exact text of an error message or prescribes a specific API call, replace
+  with a reference to the acceptance scenario or contract that defines it.
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/agents/smithy.reconcile-slices.prompt]: Fixed in b6a9395: Expanded the remediation text to explicitly cover function and type signatures: "prescribes a specific API call, or pastes a function or type signature".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt [hunk 6]
```diff
@@ -277,32 +282,105 @@ Each task is dispatched to a **fresh sub-agent** (smithy-implement) with no
...
+exact stub configurations, and exact test modifications. All of this drifts
+between planning and implementation.
+
+**GOOD** — behavioral, scannable, referencing the spec:
+
+> - [ ] **Extend `checkSpawnDependencies()` to validate all spawn preconditions**
+>
+>   Add a `baseImage` parameter to the function in `src/deps.ts`. Expand it to
+>   validate all preconditions from US2 acceptance scenarios 2.1–2.5, returning
+>   the first failure encountered.
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt:350]: Not addressed: Naming a parameter in the behavioral description ("Add a `baseImage` parameter") is describing the required API surface change — what the function should accept — not prescribing a type signature. The prohibition targets copy-pasting full signatures like `checkSpawnDependencies(baseImage: string): { ok: boolean; error?: string }`. Removing the parameter name would make the example too vague to illustrate the structured format effectively. The contracts already define this parameter, so the task is referencing the contract-level behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt [hunk 7]
```diff
@@ -277,32 +282,105 @@ Each task is dispatched to a **fresh sub-agent** (smithy-implement) with no
...
+  <2–3 sentence behavioral description. Name the target file/module and
+   the outcome. Reference acceptance scenarios by ID (e.g., "AS 2.1").>
+
+  _Acceptance criteria:_
+  - <observable invariant or behavior to verify>
+  - <another criterion>
+  - ...
+```
+
+- **Title**: imperative verb phrase, max 12 words. Scannable at a glance.
```
Comment (JMBattista) [inline src/templates/agent-skills/commands/smithy.cut.prompt]: The `- [ ] **<Imperative verb phrase...>**` is clearly the `**Title**` however the term 'title' isn't referenced in the example. Unlike for description and acceptance criteria. This could lead to agent confusion.

**Review by JMBattista — COMMENTED** [review]
Addressed review feedback in b6a9395 and 2453514.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/templates/agent-skills/commands/smithy.cut.prompt (hunk 7 above)
Comment (JMBattista) [reply src/templates/agent-skills/commands/smithy.cut.prompt]: Fixed in 2453514: All three template placeholders are now explicitly labeled — `<Title — imperative verb phrase, max 12 words>`, `<Description — 2–3 sentences...>`, and `_Acceptance criteria:_`. Applied consistently across cut.prompt, slice.prompt, and reconcile-slices.prompt.

---
_Generated by [Claude Code](https://claude.ai/code)_
